File: vcx/wrappers/java/demo/src/main/java/Alice.java
import com.evernym.sdk.vcx.connection.ConnectionApi;
import com.evernym.sdk.vcx.credential.CredentialApi;
import com.evernym.sdk.vcx.proof.DisclosedProofApi;
import com.evernym.sdk.vcx.utils.UtilsApi;
import com.evernym.sdk.vcx.vcx.VcxApi;

import com.jayway.jsonpath.JsonPath;

import org.apache.commons.cli.CommandLine;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Scanner;
import java.util.logging.Logger;

import utils.Common;
import static utils.Common.prettyJson;
import static utils.State.StateType;

public class Alice {
    // get logger for demo - INFO configured
    static final Logger logger = Common.getDemoLogger();

    public static void main(String[] args) throws Exception {
        // Library logger setup - ERROR|WARN|INFO|DEBUG|TRACE
        Common.setLibraryLogger("ERROR");

        CommandLine options = Common.getCommandLine(args);
        if (options == null) System.exit(0);

        logger.info("#0 Initialize");
        Common.loadNullPayPlugin();

        // static configuration
        long utime = System.currentTimeMillis() / 1000;
        String provisionConfig = JsonPath.parse("{" +
                "  agency_url: 'http://localhost:8080'," +
                "  agency_did: 'VsKV7grR1BUE29mG2Fm2kX'," +
                "  agency_verkey: 'Hezce2UWMZ3wUhVkh2LfKSs8nDzWwzs2Win7EzNN3YaR'," +
                "  wallet_name: 'node_vcx_demo_alice_wallet_" + utime + "'," +
                "  wallet_key: '123'," +
                "  payment_method: 'null'," +
                "  enterprise_seed: '000000000000000000000000Trustee1'" +
                "}").jsonString();

        // Communication method. aries.
        provisionConfig = JsonPath.parse(provisionConfig).put("$", "protocol_type", "3.0").jsonString();
        logger.info("Running with Aries VCX Enabled! Make sure VCX agency is configured to use protocol_type 3.0");

        if (options.hasOption("postgres")) {
            Common.loadPostgresPlugin();
            provisionConfig = JsonPath.parse(provisionConfig).put("$", "wallet_type", "postgres_storage")
                    .put("$", "storage_config", "{\"url\":\"localhost:5432\"}")
                    .put("$", "storage_credentials", "{\"account\":\"postgres\",\"password\":\"mysecretpassword\"," +
                            "\"admin_account\":\"postgres\",\"admin_password\":\"mysecretpassword\"}").jsonString();
            logger.info("Running with PostreSQL wallet enabled! Config = " + JsonPath.read(provisionConfig, "$.storage_config"));
        } else {
            logger.info("Running with builtin wallet.");
        }

        logger.info("#8 Provision an agent and wallet, get back configuration details: \n" + prettyJson(provisionConfig));
        String vcxConfig = UtilsApi.vcxProvisionAgent(provisionConfig);

        vcxConfig = JsonPath.parse(vcxConfig).put("$", "institution_name", "alice")
                .put("$", "institution_logo_url", "http://robohash.org/345")
                .put("$", "protocol_version", "2")
                .put("$", "genesis_path", System.getProperty("user.dir") + "/genesis.txn").jsonString();
        logger.info("#9 Initialize libvcx with new configuration\n" + prettyJson(vcxConfig));
        VcxApi.vcxInitWithConfig(vcxConfig).get();

        logger.info("Input faber invitation details\nEnter your invite details:");
        Scanner sc = new Scanner(System.in);
        String details = sc.nextLine();

        logger.info("#10 Convert to valid json and string and create a connection to faber");
        int connectionHandle = ConnectionApi.vcxCreateConnectionWithInvite("faber", details).get();
        ConnectionApi.vcxConnectionConnect(connectionHandle, "{\"use_public_did\": true}").get();
        int connectionState = ConnectionApi.vcxConnectionUpdateState(connectionHandle).get();
        while (connectionState != StateType.Accepted) {
            Thread.sleep(2000);
            connectionState = ConnectionApi.vcxConnectionUpdateState(connectionHandle).get();
        }

        logger.info("#11 Wait for faber to issue a credential offer");
        Thread.sleep(5000);
        String offers = CredentialApi.credentialGetOffers(connectionHandle).get();
        logger.info("Alice found " + JsonPath.read(offers, "$.length()") + " credential offers.");
        String credentialOffer = JsonPath.parse((List)JsonPath.read(offers, "$.[0]")).jsonString();
        logger.info("credential offer:\n" + prettyJson(credentialOffer));

        // Create a credential object from the credential offer
        int credentialHandle = CredentialApi.credentialCreateWithOffer("credential", credentialOffer).get();

        logger.info("#15 After receiving credential offer, send credential request");
        CredentialApi.credentialSendRequest(credentialHandle, connectionHandle, 0).get();

        logger.info("#16 Poll agency and accept credential from faber");
        int credentialState = CredentialApi.credentialUpdateState(credentialHandle).get();
        while (credentialState != StateType.Accepted) {
            Thread.sleep(2000);
            credentialState = CredentialApi.credentialUpdateState(credentialHandle).get();
        }

        logger.info("#22 Poll agency for a proof request");
        String requests = DisclosedProofApi.proofGetRequests(connectionHandle).get();
        while (JsonPath.read(requests, "$.length()").equals("0")) {
            Thread.sleep(2000);
            requests = DisclosedProofApi.proofGetRequests(connectionHandle).get();
        }
        logger.info("Alice found " + JsonPath.read(requests, "$.length()") + " proof requests.");
        String proofRequest = JsonPath.parse((LinkedHashMap)JsonPath.read(requests, "$.[0]")).jsonString();
        logger.info("proof request:\n" + prettyJson(proofRequest));

        logger.info("#23 Create a Disclosed proof object from proof request");
        int proofHandle = DisclosedProofApi.proofCreateWithRequest("proof", proofRequest).get();

        logger.info("#24 Query for credentials in the wallet that satisfy the proof request");
        String credentials = DisclosedProofApi.proofRetrieveCredentials(proofHandle).get();

        LinkedHashMap<String, Object> attrs = JsonPath.read(credentials, "$.attrs");
        for(String key : attrs.keySet()){
            String attr = JsonPath.parse((LinkedHashMap)JsonPath.read(credentials, "$.attrs." + key + ".[0]")).jsonString();
            credentials = JsonPath.parse(credentials).set("$.attrs." + key, JsonPath.parse("{\"credential\":"+ attr + "}").json()).jsonString();
        }

        logger.info("#25 Generate the proof");
        DisclosedProofApi.proofGenerate(proofHandle, credentials, "{}").get();

        logger.info("#26 Send the proof to faber");
        DisclosedProofApi.proofSend(proofHandle, connectionHandle).get();

        logger.info("#27 Wait for Faber to receive the proof");
        int proofState = DisclosedProofApi.proofUpdateState(proofHandle).get();
        while (proofState != StateType.Accepted) {
            Thread.sleep(2000);
            proofState = DisclosedProofApi.proofUpdateState(proofHandle).get();
        }
        logger.info("Faber received the proof");

        System.exit(0);
    }
}

File: vcx/wrappers/java/demo/src/main/kotlin/com/usb/sovrin/Alice.kt
package com.usb.sovrin


import ch.qos.logback.classic.Level
import ch.qos.logback.classic.Logger
import com.evernym.sdk.vcx.ErrorCode
import com.evernym.sdk.vcx.LibVcx
import com.evernym.sdk.vcx.VcxException
import com.evernym.sdk.vcx.connection.ConnectionApi
import com.evernym.sdk.vcx.credential.CredentialApi
import com.evernym.sdk.vcx.proof.DisclosedProofApi
import com.evernym.sdk.vcx.proof.ProofApi
import com.evernym.sdk.vcx.vcx.VcxApi
import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.reflect.TypeToken
import com.sun.jna.Library
import com.sun.jna.Native
import org.slf4j.LoggerFactory
import java.io.File
import java.util.*
import java.util.concurrent.ExecutionException


fun main(args: Array<String>) {




    var defaultLevel: Level = Level.OFF

    if(args.isNotEmpty() && args.size==2 && args[0] == "loglevel"){
        defaultLevel = Level.toLevel(args[1].toUpperCase())
    }
    val root = LoggerFactory.getLogger(Logger.ROOT_LOGGER_NAME) as Logger
    root.setLevel(defaultLevel)

    var libvcx = "llib/libvcx.so"

    if(System.getProperty("os.name") == "Mac OS X"){
        libvcx = "lib/libvcx.dylib"
    }

    LibVcx.init(File(libvcx))


    var libnullpay = "lib/libnullpay.so"

    if(System.getProperty("os.name") == "Mac OS X"){
        libnullpay = "lib/libnullpay.dylib"
    }

    /**Initialize Lib Null Pay library for the payment APIs. Currently there is no java wrapper available for libnullpay So we have to manually
     * load the library. This code written base off of LibVCX.init()
     * com.sun.jna is used to bridge the gap between java and OS specific DLL(libraries)
     **/
    Native.loadLibrary(File(libnullpay).absolutePath,PAYMENT_API::class.java)
            .nullpay_init()

    try {

        println("####################################################################")
        println("# 7 Provision an agent and wallet, get back configuration details")
        println("####################################################################")


        //Initialize the vcx with provision agent. This is only needed for local network
        //For Test network use vcxInitWithConfig directly
        var vcxconfig = vcxProvisionAgent("lib/alice-provisionconfig.json"
                ,"alice","http://robohash.org/456",
                "lib/localpoolconfig.json" )


        println("####################################################################")
        println("# 8 Initialize libvcx with new configuration")
        println("####################################################################")

        //Initialize vcx with the vcxconfig received from the provision agent initialization
        VcxApi.vcxInitWithConfig(vcxconfig).get()
        println("=====Configuration Initialized====")



        println("####################################################################")
        println("# 9 Input faber.py invitation details")
        println("####################################################################")

        val connectionDetail = readLine()!!

        println("####################################################################")
        println("# 10 Convert to valid json and string and create a connection to faber")
        println("####################################################################")

        var conn_to_faber = ConnectionApi.vcxCreateConnectionWithInvite("faber",connectionDetail).get()

        ConnectionApi.vcxConnectionConnect(conn_to_faber,"{\"use_public_did\": true}").get()

        ConnectionApi.vcxConnectionUpdateState(conn_to_faber).get()

        println("####################################################################")
        println("# 11 Wait for faber.py to issue a credential offer")
        println("####################################################################")

        Thread.sleep(30000)

        var offer = CredentialApi.credentialGetOffers(conn_to_faber).get()
        println("=====Offer recevied from Faber ${offer}")


        val listType = object : TypeToken<ArrayList<ArrayList<CredentialOffer>>>() {}.getType()
        val offers: List<List<CredentialOffer>> = Gson().fromJson(offer, listType)


        println("=====Creating credential with offer")


        val credOffer = CredentialApi.credentialCreateWithOffer("credential",Gson().toJson(offers.get(0))).get()


        println("####################################################################")
        println("# 15 After receiving credential offer, send credential request")
        println("####################################################################")

        CredentialApi.credentialSendRequest(credOffer,conn_to_faber,0).get()

        println("####################################################################")
        println("# 16 Poll agency and accept credential offer from faber")
        println("####################################################################")
        CredentialApi.credentialUpdateState(credOffer).get()

        var credReqState = CredentialApi.credentialGetState (credOffer).get()


        while(credReqState!=IssuerCredentialState.Accepted.code){

            Thread.sleep(2000)
            CredentialApi.credentialUpdateState(credOffer).get()
            credReqState = CredentialApi.credentialGetState (credOffer).get()


        }
        var cred = CredentialApi.getCredential(credOffer).get()
        println("======CREDENTIAL IS: ${cred}")

        println("####################################################################")
        println("# 22 Poll agency for a proof request")
        println("####################################################################")
        var proofRequests = DisclosedProofApi.proofGetRequests(conn_to_faber).get()


        println(proofRequests)

        val proofType = object : TypeToken<ArrayList<Map<String, Any>>>() {}.getType()
        val proofs: MutableList<MutableMap<String,Any>> = Gson().fromJson(proofRequests, proofType)

        val topic:MutableMap<String,Any> =  proofs.get(0).get("@topic") as MutableMap<String,Any>
        topic.put("mid", (topic.get("mid") as Double).toInt())
        topic.put("tid", (topic.get("tid") as Double).toInt())



        var proofStr = GsonBuilder().serializeNulls()
                .create().toJson(proofs.get(0))

        println("proof requests received ${proofStr}")



        println("####################################################################")
        println("# 23 Create a Disclosed proof object from proof request")
        println("####################################################################")
        val newProof = DisclosedProofApi.proofCreateWithRequest("proof",proofStr).get()

        println("####################################################################")
        println("# 24 Query for credentials in the wallet that satisfy the proof request")
        println("####################################################################")
        val credentialsJson = DisclosedProofApi.proofRetrieveCredentials(newProof).get()

        println("====Credentials in Wallets are ${credentialsJson}")

        val credType = object : TypeToken<Map<String, Any>>() {}.getType()
        val credentials: MutableMap<String,Any> = Gson().fromJson(credentialsJson, credType)

        var proof = credentials.map {


            // println(it.key)
            var attrs = it.value as Map<String,Any>

            var credentialAttrs =attrs.map {

                var credDetail = it.value as List<Any>
                //println(y.get(0))

                Pair(it.key, mapOf(Pair("credential",credDetail.get(0))))
            }.toMap()

            // println (credentialAttrs)
            Pair(it.key,credentialAttrs)
        }.toMap()

        val generatedProof =  Gson().toJson(proof)


        println("####################################################################")
        println("# 25 Generate the proof")
        println("####################################################################")

        DisclosedProofApi.proofGenerate(newProof,generatedProof,"{}").get()

        println("####################################################################")
        println("# 26 Send the proof to faber")
        println("####################################################################")


        DisclosedProofApi.proofSend(newProof,conn_to_faber).get()



    }catch(e: ExecutionException){


        var cause = e.cause
        if(cause !=null && cause is VcxException){

            println("======${ErrorCode.valueOf(cause.sdkErrorCode).name}")

        }
    }finally {

        VcxApi.vcxShutdown(false)

    }
}
